src/NHibernate/Linq/Visitors/QueryModelVisitor.cs

```diff
@@ -128,16 +133,148 @@ static QueryModelVisitor()
 			ResultOperatorMap.Add<CastResultOperator, ProcessCast>();
 		}
 
-		private QueryModelVisitor(VisitorParameters visitorParameters, bool root, QueryModel queryModel)
+		private QueryModelVisitor(VisitorParameters visitorParameters, bool root, QueryModel queryModel,
+			NhLinqExpressionReturnType? rootReturnType)
 		{
 			VisitorParameters = visitorParameters;
 			Model = queryModel;
+			_rootReturnType = root ? rootReturnType : null;
 			_hqlTree = new IntermediateHqlTree(root);
 		}
 
 		private void Visit()
 		{
 			VisitQueryModel(Model);
+			AddAdditionalPostExecuteTransformer();
+		}
+
+		private void AddAdditionalPostExecuteTransformer()
+		{
+			if (_rootReturnType == NhLinqExpressionReturnType.Scalar && Model.ResultTypeOverride != null)
+			{
+				// NH-3850: handle polymorphic scalar results aggregation
+				switch ((NhExpressionType)Model.SelectClause.Selector.NodeType)
+				{
+					case NhExpressionType.Average:
+						// Polymorphic case complex to handle and not implemented. (HQL query must be reshaped for adding
+						// additional data to allow a meaningful overall average computation.)
+						// Leaving it untouched for allowing non polymorphic cases to work.
+						break;
+					case NhExpressionType.Max:
+						AddPostExecuteTransformerForResultAggregate(ReflectionCache.QueryableMethods.MaxDefinition);
```

Comment:
The `IntermediateHqlTree._postExecuteTransformers` is a list of `LambdaExpression` which get merged together then compiled and called on `IQueryable<T>`. As if they should be chainable, as if we it were useful and possible to add many of them, though the usages I have currently seen for them were all incompatible with chaining (they all yield another type than their input).

Yes that does not make sens to me either, that overly complicates things for defining the post execute transform (messing with expressions for runtime execution instead of just supplying directly an executable delegate), but there are no comments as for why such a design decision, and I have not decided to redesign this.

Redesign those transformers should imply considering the design of the list transformer, which works the same way and is way more used. But maybe this one has some cases where it could be translated to DB, though I doubt.